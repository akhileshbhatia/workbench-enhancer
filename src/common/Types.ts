export type ChromeStorageQueryData = {
  data: string
};

export type TimeDetailsMap = Map<number, ChromeStorageQueryData>;

export type QueryDataMap = Map<string, TimeDetailsMap>;
