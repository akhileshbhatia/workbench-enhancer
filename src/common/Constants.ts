export const extensionStateKey = 'extension_states';
export const appId = 'wb-enhancer-app';
export const mainBlockId = 'mainBlock';
