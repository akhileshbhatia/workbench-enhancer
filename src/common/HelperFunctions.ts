import { QueryDataMap, ChromeStorageQueryData, TimeDetailsMap } from './Types';

export function getDataFromChromeStorage(key: string): Promise<Record<string, unknown>> {
  return new Promise((resolve, reject) => {
    try {
      chrome.storage.local.get(key, (data) => {
        key === null ? resolve(data) : resolve(data[key]);
      });
    } catch (err) {
      reject(err);
    }
  });
}

export function clearStorage(): Promise<void> {
  return new Promise((resolve, reject) => {
    try {
      chrome.storage.local.clear(() => resolve());
    } catch (err) {
      reject(err);
    }
  });
}

export function setDataToChromeStorage(key: string, data: any): Promise<void> {
  return new Promise((resolve, reject) => {
    try {
      chrome.storage.local.set({ [key]: data }, () => {
        if (chrome.runtime.lastError) {
          reject('Error in setting data to chrome storage');
        }
        resolve();
      });
    } catch (err) {
      reject(err);
    }
  });
}

export function serializeMap<keyType, valueType>(map: Map<keyType, valueType>): string {
  return JSON.stringify(Array.from(map.entries()));
}

export function deserializeToMap<keyType, valueType>(data: string): Map<keyType, valueType> {
  return new Map(JSON.parse(data));
}

export function deserializeData(data: string): { output: QueryDataMap } {
  const finalMap = new Map();
  if (data) {
    const dateMap = deserializeToMap<string, string>(data);
    for (const [date, info] of dateMap.entries()) {
      const newValue = deserializeToMap<number, ChromeStorageQueryData>(info);
      finalMap.set(date, newValue);
    }
  }
  return { output: finalMap };
}

export function getHoursAndMinsFromTimestamp(timestamp: number): string {
  const date = new Date(timestamp * 1000);
  // Pad 0 to hours and mins in case they are single digit
  const hours = String(date.getHours()).padStart(2, '0');
  const mins = String(date.getMinutes()).padStart(2, '0');
  return `${hours}:${mins}`;
}

export function getFormattedDateAndTimestamp(): {
  formattedDate: string,
  timestamp: number
} {
  const todaysDate = new Date();
  const year = new Intl.DateTimeFormat('en', { year: 'numeric' }).format(todaysDate);
  const month = new Intl.DateTimeFormat('en', { month: 'short' }).format(todaysDate);
  const day = new Intl.DateTimeFormat('en', { day: '2-digit' }).format(todaysDate);
  return {
    formattedDate: `${day} ${month} ${year}`,
    timestamp: Math.round(<number><unknown>todaysDate / 1000)
  };
}

export function updateTextArea(value: string): void {
  document.querySelector('textarea').value = value;
}

export async function formatDataAndAddtoStorage(
  dataArray: [string, TimeDetailsMap][],
  currentPathName: string
): Promise<void> {
  const dataToStore = new Map<string, string>();
  for (const [date, details] of dataArray) { // Serialize data again before storing
    dataToStore.set(date, serializeMap<number, Record<string, unknown>>(details));
  }
  await setDataToChromeStorage(currentPathName, serializeMap<string, string>(dataToStore));
}
