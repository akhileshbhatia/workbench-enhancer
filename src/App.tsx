import React, { ReactElement, useState, ChangeEvent } from 'react';
import './app.scss';
import {
  Drawer,
  IconButton,
  makeStyles,
  Theme,
  TextField,
  InputAdornment
} from '@material-ui/core';
import ChevronRight from '@material-ui/icons/ChevronRight';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import SearchIcon from '@material-ui/icons/Search';

import clsx from 'clsx';
import { updateExtensionState, deleteFromStorage } from './StorageServices';
import { QueryAccordion } from './QueryAccordion';
import { QueryDataMap } from './common/Types';

const drawerWidth = 240;

const useStyles = makeStyles((theme: Theme) => ({
  root: {
    display: 'flex',
    width: '100%'
  },
  menuButton: {
    marginRight: theme.spacing(2)
  },
  hide: {
    display: 'none'
  },
  drawer: {
    width: drawerWidth,
    flexShrink: 0
  },
  drawerPaper: {
    width: drawerWidth
  },
  drawerHeader: {
    display: 'flex',
    alignItems: 'center',
    padding: theme.spacing(0, 1),
    ...theme.mixins.toolbar,
    justifyContent: 'flex-end'
  },
  inputField: {
    margin: theme.spacing(1),
    width: '90%',
    height: '70px'
  }
}));

export type AppProps = {
  output: QueryDataMap,
  defaultDrawerState: boolean,
  currentPathName: string
};

export function App(props: AppProps): ReactElement {
  const { output, defaultDrawerState, currentPathName } = props;
  const classes: Record<string, string> = useStyles();
  const [drawerOpen, setDrawerOpen] = useState(defaultDrawerState);
  const [searchTerm, setSearchTerm] = useState('');
  const [allData, setAllData] = useState(output);

  const updateDrawerState = async (newState: boolean) => {
    setDrawerOpen(newState);
    await updateExtensionState(currentPathName, newState);
  };

  const handleSearchTermUpdate = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleDelete = async (timestamp: number, dateToDeleteFrom: string) => {
    const updatedData = await deleteFromStorage(timestamp, dateToDeleteFrom, currentPathName, allData);
    setAllData(() => new Map([...updatedData]));
  };

  return (
    <div className={classes.root}>
      <IconButton
        color="inherit"
        aria-label="open drawer"
        data-testid="open-drawer-button"
        onClick={() => updateDrawerState(true)}
        edge="end"
        className={clsx(classes.menuButton, drawerOpen && classes.hide)}
      >
        <ChevronRight />
      </IconButton>
      <Drawer
        className={classes.drawer}
        variant="persistent"
        anchor="left"
        open={drawerOpen}
        classes={{
          paper: classes.drawerPaper
        }}
      >
        <div className={classes.drawerHeader}>
          <IconButton onClick={() => updateDrawerState(false)} data-testid="close-drawer-button">
            <ChevronLeftIcon />
          </IconButton>
        </div>
        <div data-testid="no-data-msg" hidden={allData.size !== 0}>
          <h2>Please add queries/searches to see them here!</h2>
        </div>
        <div data-testid="data-div" hidden={allData.size === 0}>
          <TextField
            variant="outlined"
            className={classes.inputField}
            value={searchTerm}
            onChange={handleSearchTermUpdate}
            data-testid="search-text-field"
            InputProps={{
              startAdornment: (
                <InputAdornment position="start">
                  <SearchIcon />
                </InputAdornment>
              )
            }}
          />
        </div>
        <div>
          {
            [...allData.keys()].map(date => {
              const entries = [...allData.get(date).entries()];
              const props = { date, entries, searchTerm, handleDelete };
              return <QueryAccordion key={date} {...props} />;
            })
          }
        </div>
      </Drawer>
    </div >
  );
}
