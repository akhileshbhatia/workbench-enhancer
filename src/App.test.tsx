import React from 'react';
import { AppProps, App } from './App';
import { RenderResult, render, fireEvent, getByTestId, waitFor } from '@testing-library/react';
import { QueryDataMap, TimeDetailsMap } from './common/Types';
import { getRandomTimestamp } from './test/HelperFunctions';

describe('App', () => {
  let component: RenderResult;
  let queryDataMap: QueryDataMap;
  let timeDetailsMap: TimeDetailsMap;

  beforeEach(() => {
    jest.useFakeTimers();
    queryDataMap = new Map() as QueryDataMap;
    timeDetailsMap = new Map() as TimeDetailsMap;
  });

  const init = (dataArray: string[] = [], defaultDrawerState = true, currentPathName = 'query'): void => {
    dataArray.map(data => {
      timeDetailsMap.set(getRandomTimestamp(), { data });
    });
    if (timeDetailsMap.size) { // Only add to queryDataMap if there is atleast one data
      queryDataMap.set('12 March 2021', timeDetailsMap);
    }
    const props: AppProps = {
      output: queryDataMap,
      defaultDrawerState,
      currentPathName
    };
    component = render(<App {...props} />);
  };

  const getDrawerDivVisibility = (): string => {
    const drawer = component.container.getElementsByClassName('MuiDrawer-root')[0].firstElementChild;
    return window.getComputedStyle(drawer).visibility;
  };

  it('should match snapshot', () => {
    init();
    expect(component).toMatchSnapshot();
  });

  it('should hide the data div when there is no data', () => {
    init();
    expect(getByTestId(component.container, 'no-data-msg').hasAttribute('hidden')).toBeFalsy();
    expect(getByTestId(component.container, 'data-div').hasAttribute('hidden')).toBeTruthy();
  });

  it('should open the drawer by default', () => {
    init();
    expect(getDrawerDivVisibility()).toBe('visible');
  });

  it('should close the drawer by default', () => {
    init(undefined, false);
    expect(getDrawerDivVisibility()).toBe('hidden');
  });

  it('should toggle the drawer when button is clicked', async () => {
    const testPathName = 'test-path';
    init(undefined, undefined, testPathName);
    fireEvent.click(getByTestId(component.container, 'close-drawer-button'));
    await waitFor(() => {
      expect(getDrawerDivVisibility()).toBe('hidden');
    });
    fireEvent.click(getByTestId(component.container, 'open-drawer-button'));
    await waitFor(() => {
      expect(getDrawerDivVisibility()).toBe('visible');
    });
  });
});
