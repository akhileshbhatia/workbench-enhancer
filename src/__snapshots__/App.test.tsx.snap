// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-root-1"
      >
        <button
          aria-label="open drawer"
          class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 makeStyles-hide-3 MuiIconButton-colorInherit MuiIconButton-edgeEnd"
          data-testid="open-drawer-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiDrawer-root MuiDrawer-docked makeStyles-drawer-4"
        >
          <div
            class="MuiPaper-root MuiDrawer-paper makeStyles-drawerPaper-5 MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
          >
            <div
              class="makeStyles-drawerHeader-6"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="close-drawer-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              data-testid="no-data-msg"
            >
              <h2>
                Please add queries/searches to see them here!
              </h2>
            </div>
            <div
              data-testid="data-div"
              hidden=""
            >
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-inputField-7"
                data-testid="search-text-field"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
                >
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                      />
                    </svg>
                  </div>
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                    style="padding-left: 8px;"
                  >
                    <legend
                      class="PrivateNotchedOutline-legend-9"
                      style="width: 0.01px;"
                    >
                      <span>
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-root-1"
    >
      <button
        aria-label="open drawer"
        class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 makeStyles-hide-3 MuiIconButton-colorInherit MuiIconButton-edgeEnd"
        data-testid="open-drawer-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiDrawer-root MuiDrawer-docked makeStyles-drawer-4"
      >
        <div
          class="MuiPaper-root MuiDrawer-paper makeStyles-drawerPaper-5 MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
        >
          <div
            class="makeStyles-drawerHeader-6"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              data-testid="close-drawer-button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            data-testid="no-data-msg"
          >
            <h2>
              Please add queries/searches to see them here!
            </h2>
          </div>
          <div
            data-testid="data-div"
            hidden=""
          >
            <div
              class="MuiFormControl-root MuiTextField-root makeStyles-inputField-7"
              data-testid="search-text-field"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                    />
                  </svg>
                </div>
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                  style="padding-left: 8px;"
                >
                  <legend
                    class="PrivateNotchedOutline-legend-9"
                    style="width: 0.01px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
